Функциональность «Общие фильмы»

Phronima6 opened the pull request:
Добавить:
- функциональность «Общие фильмы»;
- диаграмму базы данных с пояснениями.

Commit:
Добавить:
- функциональность «Общие фильмы»;
- диаграмму базы данных с пояснениями.

LadyMoonMare commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmService.java:
```diff
@@ -20,4 +20,11 @@ public interface FilmService {
+    List<Film> getCommonFilms(Integer userId, Integer friendId);
+
+    // Проверяем есть ли такой пользователь
+    void checkUserById(Integer userId);
```
Это метод внутренней валидации для FilmServiceImpl. В дальнейшем нигде кроме этого файла не используется, поэтому можно не добавлять его в интерфейс. На усмотрение.

zolotyh-dk replied:
> можно не добавлять его в интерфейс

Согласен. По крайней мере не в интерфейс `FilmService`

LadyMoonMare reviewed (commented):
Отличная работа!
От меня одно замечание на твоё усмотрение.

zolotyh-dk commented on diagramFilmorate.svg:
На мой взгляд, промежуточные таблицы feed_event_type и feed_operation не нужны. Промежуточная таблица нужна когда связь "многие ко многим". В нашем случае, каждая запись в таблице feed имеет только один eventType и одну operation. Следовательно, связь между feed и event_type и с operations является простой ссылочной связью через внешний ключ. Добавление промежуточных таблиц здесь усложнит структуру без необходимости, так как мы просто ссылаемся на заранее определенные типы.

Предлагаю пока оставить и уточнить у наставника.

zolotyh-dk commented on src/main/java/ru/yandex/practicum/filmorate/service/film/FilmServiceImpl.java:
```diff
@@ -130,4 +134,27 @@ public List<Film> getFilmsByDirector(int directorId, String sortBy) {
-}
+
+    @Override
+    public List<Film> getCommonFilms(Integer userId, Integer friendId) {
```
Есть проверка, что такие юзеры есть в БД, но нет проверки, что они действительно друзья. Сейчас любой пользователь может запросить общие фильмы с любым другим пользователем

zolotyh-dk reviewed (approved):
Хорошая работа, Артём! Добавил несколько комментариев в местах, на которые стоит обратить внимаение.

Phronima6 commented:
Спасибо ребята! Исправлю код.

Commit:
Привести в соответствие с замечаниями:
- убрать из FilmServiceImpl метод checkUserById, откорректировать проверку в методе getCommonFilms;
- добавить в FilmServiceImpl, в методе getCommonFilms проверку на подтверждённую дружбу.

LadyMoonMare reviewed (approved):
Спасибо! Работу принимаю)
